### PR TITLE
[research feature] Introduce INT4 support at the algorithm level

### PR DESCRIPTION
**Pull request by wenhuach21**
## Description

Please briefly describe your main changes, the motivation.

- [ ] delay rotation to save ram/vram
- [ ] inference
- [ ] automodel and automodelforcacuallm should be different

### CODE REVIEW

File: auto_round/compressors/base.py [hunk 1]
```diff
@@ -557,18 +557,33 @@ def __init__(
...
+            # from auto_round.experimental.utils import normalize_hadamard_config
+
+            #
+            # _hcfg = normalize_hadamard_config(hadamard_config)
+            # if not isinstance(_hcfg, HadamardConfig):
+            #     _hcfg = HadamardConfig(**_hcfg)
+            # group_size = _hcfg.block_size  # e.g. 32 for MXFP4
+
+            self.model.to("cuda")
+            apply_hadamard_rotation(self.model, group_size=self.group_size, allow_online_hadamard=False)
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]:
```suggestion
            apply_hadamard_rotation(self.model, group_size=self.group_size, allow_online_hadamard=True)
```

**Comment by wenhuach21** [comment]
please have a review

File: auto_round/compressors/utils.py [hunk 2]
```diff
@@ -107,14 +107,24 @@ def is_wint_a16(ar: Callable) -> bool:
...
     if is_wint8aint8(ar_or_format):
         return True
     return False
 
 
+def is_wint4aint4(ar_or_scheme: Union[str, Callable]):
+    if isinstance(ar_or_scheme, str):
+        return "int4" in ar_or_scheme.lower()
+    elif (
+        "int4" in ar_or_scheme.act_data_type or ("int" in ar_or_scheme.act_data_type and ar_or_scheme.act_bits == 4)
```
Comment (hshen14) [inline auto_round/compressors/utils.py:122]: It assumes act_data_type is always non-None. Maybe we can add assertion here.

File: docs/environments.md [hunk 3]
```diff
@@ -69,6 +69,26 @@ export AR_DISABLE_OFFLOAD=1
 export AR_DISABLE_DATASET_SUBPROCESS=true
 ```
 
+### AR_ACT_SCALE
+- **Description**: Controls the scaling factor applied to activation min/max values during activation quantization. A value less than 1.0 shrinks the clipping range, which can reduce outlier impact.
+- **Default**: `1.0`
+- **Valid Values**: Any float, e.g. `0.8`, `0.9`, `1.0`
```
Comment (hshen14) [inline docs/environments.md]: Inconsistent with CN version? 1.0 vs. 0.9

Also the example 0.9 vs. 0.8

File: auto_round/inference/convert_model.py [hunk 4]
```diff
@@ -680,19 +680,19 @@ def convert_hf_model(model: nn.Module, target_device: str = "cpu") -> tuple[nn.M
...
         # apply forward hook
-        act_hadamard_config = HadamardConfig(
-            block_size=hadamard_config["block_size"],
-            hadamard_type=hadamard_config["hadamard_type"],
+        act_rotation_config = RotationConfig(
+            block_size=rotation_config["block_size"],
+            hadamard_type=rotation_config["hadamard_type"],
         )  # apply to activation
-        model = apply_hadamard_transform(
+        model = apply_rotation_transform(
```
Comment (hshen14) [inline auto_round/inference/convert_model.py]: should use tuple (model, _)?

File: auto_round/compressors/utils.py (hunk 2 above)
Comment (wenhuach21) [reply auto_round/compressors/utils.py:122]: That's true , `act_data_type` will be overridden by the weight data type if it’s not explicitly set. I can refine this behavior.

**Review by lkk12014402 — APPROVED** [review]
LGTM

**Comment by lkk12014402** [comment]
By the way, is the implementation of Quarot consistent with community implementations like llm-compressor? Have we compared the accuracy?

**Comment by wenhuach21** [comment]
> By the way, is the implementation of Quarot consistent with community implementations like llm-compressor? Have we compared the accuracy?

I’m not entirely sure. The Quarot accuracy is evaluated using their own code, while we focus specifically on the Hadamard transformation, where I believe our implementation is consistent with theirs. One key difference is that they use a random Hadamard transformation, whereas we use a fixed Hadamard, which we’ve found to perform better.

**Comment by wenhuach21** [comment]
[0374bd2](https://github.com/intel/auto-round/pull/1641/commits/0374bd20d3e1b1510e3f7350720989f4a25510f4) accuracy good
56505e8
03c861b has issues
